### PR TITLE
Add api for obol

### PR DESCRIPTION
**Pull request by evgeny-stakewise**

### CODE REVIEW

File: src/validators/endpoints.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import asyncio
```
Comment (cyc60) [inline src/validators/endpoints.py:1]: move this file to api package?

File: src/validators/endpoints.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    pending_validator_registrations,
+    register_validators,
+)
+from src.validators.typings import Validator, ValidatorsRegistrationMode
+
+
+async def get_validators(request: Request) -> Response:
+    deposit_data = request.app.state.deposit_data
+
+    validators = await get_available_validators_for_registration(
```
Comment (cyc60) [inline src/validators/endpoints.py:27]: `check_deposit_data_root` with contact call called in `get_available_validators_for_registration`
MB refactor get_available_validators_for_registration to speed up api response.

File: src/validators/tasks.py [hunk 3]
```diff
@@ -57,22 +63,45 @@ async def process_block(self) -> None:
...
+            return
+
         await update_unused_validator_keys_metric(
             keystore=self.keystore,
             deposit_data=self.deposit_data,
         )
-        await register_validators(
-            keystore=self.keystore,
-            deposit_data=self.deposit_data,
+        if settings.validators_registration_mode == ValidatorsRegistrationMode.AUTO:
```
Comment (cyc60) [inline src/validators/tasks.py:74]: validators_registration_mode always set with no_keystore.

File: src/validators/tasks.py [hunk 4]
```diff
@@ -29,22 +30,27 @@
...
     NetworkValidator,
     Validator,
+    ValidatorsRegistrationMode,
 )
 from src.validators.utils import send_approval_requests
 
 logger = logging.getLogger(__name__)
 
 
+pending_validator_registrations: list[HexStr] = []
```
Comment (tsudmi) [inline src/validators/tasks.py:47]: Would that work if I run with `--pool-size=2`?

File: src/validators/api/endpoints.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        keystore=None, deposit_data=deposit_data
+    )
+
+    error = await _validate_approvals(approvals, validators)
+    if error is not None:
+        return JSONResponse({'error': error}, status_code=HTTP_400_BAD_REQUEST)
+
+    # There may be lag between GET and POST requests.
+    # During this time new assets may be staked into the vault.
+    # In this case validators list will be longer then approvals list.
```
Comment (tsudmi) [inline src/validators/api/endpoints.py:77]:
```suggestion
    # In this case validators list will be longer than approvals list.
```

File: src/validators/api/endpoints.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    # Filter validators by requested public keys.
+    validators = validators[: len(approvals)]
```
Comment (tsudmi) [inline src/validators/api/endpoints.py:79]: what if it's smaller? When deposits enter exit queue, they will reduce amount of ETH that can be staked

File: src/validators/api/endpoints.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        validator.exit_signature = approval.exit_signature
+
+    pending_validator_registrations.extend([v.public_key for v in validators])
+    asyncio.create_task(
+        register_and_remove_pending_validators(
+            keystore=None, deposit_data=deposit_data, validators=validators
+        )
+    )
+
+    return JSONResponse('ok')
```
Comment (tsudmi) [inline src/validators/api/endpoints.py:91]: Lets provide more standardized response like `"status": "success"`

File: src/validators/tasks.py [hunk 8]
```diff
@@ -229,12 +287,24 @@ async def create_approval_request(
...
-            validator_index=validator_index,
-            public_key=validator.public_key,
-            oracles=oracles,
-            fork=settings.network_config.SHAPELLA_FORK,
-        )
+        if keystore is None:
+            if validator.exit_signature is None:
+                raise RuntimeError('validator.exit_signature must be set')
+
+            shards = await LocalKeystore.get_exit_signature_shards_without_keystore(
```
Comment (tsudmi) [inline src/validators/tasks.py:294]: Lets generate shards the same way: for local keystore get the exit signature and shard it then

File: src/commands/start_api.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+    type=click.Choice(
+        LOG_LEVELS,
+        case_sensitive=False,
+    ),
+    default='INFO',
+    envvar='LOG_LEVEL',
+    help='The log level.',
+)
+@click.option(
+    '--pool-size',
```
[the remark this answers is not in the document]
Comment (evgeny-stakewise) [reply src/commands/start_api.py:179]: It affects only keystores loading. It is not server concurrency. May be deleted from this command args.

File: src/commands/start_api.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+)
+@click.option(
+    '--vault',
+    type=ChecksumAddress,
+    callback=validate_eth_address,
+    envvar='VAULT',
+    prompt='Enter the vault address',
+    help='Address of the vault to register validators for.',
+)
+@click.option(
```
[the remark this answers is not in the document]
Comment (evgeny-stakewise) [reply src/commands/start_api.py:137]: Not needed

File: src/commands/start_api.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+    '--remote-signer-url',
+    type=str,
+    envvar='REMOTE_SIGNER_URL',
+    help='The base URL of the remote signer, e.g. http://signer:9000',
+)
+@click.option(
```
[the remark this answers is not in the document]
Comment (evgeny-stakewise) [reply src/commands/start_api.py:143]: Not needed

File: src/validators/tasks.py (hunk 4 above)
Comment (evgeny-stakewise) [reply src/validators/tasks.py:47]: `pool-size` affects only keystores loading. It is not server concurrency. Server works in a single thread. Thread safety is not necessary for now.

File: src/validators/tasks.py (hunk 8 above)
Comment (evgeny-stakewise) [reply src/validators/tasks.py:294]: Thats the way it works

File: src/validators/tasks.py (hunk 4 above)
Comment (evgeny-stakewise) [reply src/validators/tasks.py:47]: pool-size affects only keystores loading. It is not server concurrency.

File: src/validators/api/endpoints.py (hunk 6 above)
Comment (evgeny-stakewise) [reply src/validators/api/endpoints.py:79]: We just checked vault assets are enough. See get_available_validators_for_registration, above

File: src/validators/tasks.py (hunk 4 above)
Comment (evgeny-stakewise) [reply src/validators/tasks.py:49]: I removed pool-size from start-api command args

File: src/validators/typings.py [hunk 12]
```diff
@@ -36,7 +36,7 @@ def public_keys(self) -> list[HexStr]:
 @dataclass
 class ExitSignatureShards:
     public_keys: list[HexStr]
-    exit_signatures: list[BLSSignature]
+    exit_signatures: list[HexStr]
```
Comment (evgeny-stakewise) [inline src/validators/typings.py]: I was thinking about making 2 dataclasses - one for raw signatures, another one for encrypted signatures. I ended up with the same dataclass as it was before.

File: src/validators/keystores/remote.py [hunk 13]
```diff
@@ -69,40 +65,14 @@ def __contains__(self, public_key):
...
-        )
-
-        return ExitSignatureShards(
-            public_keys=[Web3.to_hex(p) for p in public_key_shares],
-            exit_signatures=exit_signature_shares,
-        )
-
     async def get_exit_signature(
-        self, validator_index: int, public_key: HexStr, network: str, fork: ConsensusFork
+        self, validator_index: int, public_key: HexStr, fork: ConsensusFork | None = None
```
Comment (evgeny-stakewise) [inline src/validators/keystores/remote.py:69]: Keystores rely on settings. So why cant we take fork and network from settings?

File: src/config/settings.py [hunk 14]
```diff
@@ -67,6 +72,12 @@ class Settings(metaclass=Singleton):
     sentry_dsn: str
     pool_size: int | None
 
+    enable_api: bool
```
Comment (tsudmi) [inline src/config/settings.py:75]: I think we can `validators_registration_mode` for checks whether api should be enabled or not. Then we can remove this setting

File: src/validators/api/endpoints.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    pending_validator_registrations,
+    register_and_remove_pending_validators,
+)
+from src.validators.typings import Validator, ValidatorsRegistrationMode
+
+
+async def get_validators(request: Request) -> Response:
+    deposit_data = request.app.state.deposit_data
+
+    validators = await get_available_validators_for_registration(
```
Comment (tsudmi) [inline src/validators/api/endpoints.py:27]: As we pass `run_check_deposit_data_root=False`, lets make sure we check the deposit data root during API start.

File: src/validators/api/endpoints.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+    return JSONResponse(
+        [
+            {'public_key': validator.public_key, 'index': index}
+            for index, validator in enumerate(validators, next_validator_index)
+        ]
+    )
+
+
+async def approve_validators(request: Request) -> Response:
```
Comment (tsudmi) [inline src/validators/api/endpoints.py:44]: Maybe `submit_validators` or `register_validators`? The caller is not approving here. Lets rename also other `approvals` names in the code.

File: src/validators/api/endpoints.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+
+async def _validate_approvals(
+    approvals: list[ValidatorApproval], validators: list[Validator]
+) -> Any:
+    """
+    Business logic validation
+    :return: error
+    """
+    if len(validators) < len(approvals):
```
Comment (tsudmi) [inline src/validators/api/endpoints.py:101]: I think we should be able to proceed with such case as well. If there are more approvals that available validators for the registration, we can take `approvals[:len(validators)]`.

File: src/config/settings.py [hunk 18]
```diff
@@ -67,6 +72,12 @@ class Settings(metaclass=Singleton):
...
+    api_host: str
+    api_port: int
+    validators_registration_mode: ValidatorsRegistrationMode
+    skip_startup_checks: bool
```
[the remark this answers is not in the document]
Comment (evgeny-stakewise) [reply src/config/settings.py:79]: It's annoying to wait for startup-checks to complete.

File: src/validators/api/endpoints.py (hunk 15 above)
Comment (evgeny-stakewise) [reply src/validators/api/endpoints.py:27]: It is checked during startup-checks

**Review by tsudmi — APPROVED** [review]

**Pull request merged**